### PR TITLE
Adding storm control testplan

### PR DESCRIPTION
**Pull request by bchovva**
### Description of PR

Summary:  This PR adds a comprehensive Storm Control test plan to validate broadcast, multicast, and unknown unicast traffic suppression behaviour in SONiC. 

Fixes # (issue)

### Type of change

- [ ] Bug fix
- [ ] Testbed and Framework(new/improvement)
- [x] New Test case
    - [ ] Skipped for non-supported platforms
- [ ] Test case improvement

### Back port request
- [ ] 202311
- [ ] 202405
- [ ] 202411
- [ ] 202505
- [x] 202511
- [ ] 202512
- [x] 202605

#### What is the motivation for this PR?

This PR aims to introduce a standardized and automated test plan to validate these scenarios in SONiC test framework.
#### How did you do it?

Added storm control test scenarios covering:

Broadcast storm control
Unknown multicast storm control
Unknown unicast storm control

#### How did you verify/test it?
Executed test cases on T0 topology

#### Any platform specific information?
NA
#### Supported testbed topology if it's a new test case?
T0

### CODE REVIEW

File: docs/testplan/Storm-Control-test-plan.md [hunk 1]
```diff
@@ -0,0 +1,568 @@
...
+|-------|-------|
+| Function | `test_storm_control_negative_unconfigure_with_kbps` |
+| Priority | P4 |
+
+**Steps**
+1. Add broadcast storm-control on the interface.
+2. Run `config interface storm-control del <intf> broadcast 100` (extra kbps argument).
+3. Assert non-zero return code.
+4. Assert the original CONFIG_DB entry is still present (the invalid `del` must not have removed it).
+
```
Comment (bpar9) [inline docs/testplan/Storm-Control-test-plan.md:543]: The HLD states the kbps range is 0 to 100,000,000. No test exercise boundary values:
0 kbps — should this be rejected or treated as disable?
1 kbps (minimum non-zero)
100,000,000 kbps (maximum — 100 Gbps)
100,000,001 kbps (out of range — must be rejected)

File: docs/testplan/Storm-Control-test-plan.md [hunk 2]
```diff
@@ -0,0 +1,568 @@
...
+| Field | Value |
+|-------|-------|
+| Function | `test_storm_control_negative_missing_kbps` |
+| Priority | P2 |
+
+**Steps**
+1. Run `config interface storm-control add <intf> broadcast` (no kbps).
+2. Assert non-zero return code.
+
+#### TC 4.4.4 — Storm-control on PortChannel rejected
```
Comment (bpar9) [inline docs/testplan/Storm-Control-test-plan.md]: HLD section 2.3 states "Configuration is not supported on VLAN and port-channel interfaces." TC 4.4.4 tests PortChannel rejection, but there is no test for VLAN interface rejection, can this also be checked/verfied

File: docs/testplan/Storm-Control-test-plan.md [hunk 3]
```diff
@@ -0,0 +1,568 @@
...
+| Priority | P1 |
+
+**Steps**
+**Steps**
+1. Add broadcast storm-control on the interface.
+2. `config interface shutdown <intf>` -> wait -> `config interface startup <intf>` -> wait for link up.
+3. Read CONFIG_DB; assert kbps unchanged.
+4. Run the common measurement procedure (broadcast, policing expected); assert the forwarded rate is still capped at ~the ceiling.
+
+#### TC 4.3.2 — Cold / Warm / Fast reboot persistence
```
Comment (bpar9) [inline docs/testplan/Storm-Control-test-plan.md:387]: HLD section 1.1.4 specifically calls out SWSS docker warm restart as a warm boot requirement. No test restarts the swss container and verifies policers are re-programmed.

File: docs/testplan/Storm-Control-test-plan.md [hunk 4]
```diff
@@ -0,0 +1,568 @@
...
+
+Tunable constants:
+
+| Constant | Default | Purpose |
+|----------|---------|---------|
+| `STORM_CTRL_THRESHOLD_KBPS` | `1000` | Policer ceiling programmed for functional tests. |
+| `BURST_KBPS` | `3000` (3x threshold) | Paced send rate generated by PTF. |
+| `BURST_PKT_COUNT` | `3000` | Packets per burst (per packet size). |
+| `PKT_LENS` | `[512, 1518]` | Ethernet frame sizes exercised; every functional burst runs once per size. |
+| `RATE_TOLERANCE_PCT` | `10` | Forwarded-rate tolerance: when policing is expected the forwarded rate may exceed the ceiling by up to this percent; when not expected, the forwarded count must be within this percent of everything sent. |
```
Comment (bpar9) [inline docs/testplan/Storm-Control-test-plan.md]: where is this 10% coming from, is this acceptable for all asics/platforms or should this be made platform specific?

**Review by bpar9 — CHANGES_REQUESTED** [review]

File: docs/testplan/Storm-Control-test-plan.md (hunk 1 above)
Comment (bchovva) [reply docs/testplan/Storm-Control-test-plan.md:543]: Added test cases to address boundary values.
Added test case to verify out of range (100,000,001 kbps) value
0kbps - added testcase to validate 0 kpbs (ASIC depenedent)

File: docs/testplan/Storm-Control-test-plan.md (hunk 3 above)
Comment (bchovva) [reply docs/testplan/Storm-Control-test-plan.md:387]: Added testcase for SWSS docker restart

File: docs/testplan/Storm-Control-test-plan.md (hunk 2 above)
Comment (bchovva) [reply docs/testplan/Storm-Control-test-plan.md]: Added test case for VLAN rejection

File: docs/testplan/Storm-Control-test-plan.md (hunk 4 above)
Comment (bchovva) [reply docs/testplan/Storm-Control-test-plan.md]: Yes, it is platform/asic dependent

File: docs/testplan/Storm-Control-test-plan.md [hunk 5]
```diff
@@ -0,0 +1,643 @@
...
+| Priority | P1 |
+| Skip | If no PortChannel member exists on the DUT |
+
+**Steps**
+1. Locate any `PORTCHANNEL_MEMBER|<pc>|<intf>` key in CONFIG_DB.
+2. Add broadcast storm-control on the *physical member* interface.
+3. Assert the CLI succeeded.
+4. Assert the CONFIG_DB entry lives under the **physical port**, not under the PortChannel.
+
+**Pass criteria**: SONiC accepts storm-control on a LAG member port and stores the entry against the physical port only.
```
Comment (anders-nexthop) [inline docs/testplan/Storm-Control-test-plan.md:576]: If we enable storm-control on one link but not the others, does that affect hashing in any way? The result may be vendor-specific, but it would be good to have an idea of how these features will interact.

I didn't see any mention of PortChannels in the HLD, we might want to add a section there mentioning that PortChannels do not support storm-control, but that individual links can have storm-control enabled/disabled independently.

**Review by anders-nexthop — COMMENTED** [review]
One question about PortChannel interactions that was brought up during the testplan review, mentioning it here as well.

**Review by bpar9 — APPROVED** [review]

File: docs/testplan/Storm-Control-test-plan.md (hunk 5 above)
Comment (bchovva) [reply docs/testplan/Storm-Control-test-plan.md:576]: Storm-control is typically an ingress feature and should not affect PortChannel hashing directly. Hashing decisions are made independently based on the configured load-balancing algorithm, so enabling storm-control on one member link would not normally change how traffic is distributed across the bundle. 

Regarding PortChannels in HLD, this is already called out in the documentation:

HLD Section 2.3 – Limitations mentions that storm-control is not supported on PortChannels.
Test Plan TC 4.4.5 – "Storm-control on PortChannel rejected" validates this behavior.

Since storm-control is applied at the individual physical interface level, member links can have storm-control enabled or disabled independently.

**Review by anders-nexthop — APPROVED** [review]

**Comment by StormLiangMS** [comment]
@bchovva — this PR is labeled Request for 202605 branch, but the following required information is still missing:

- linked GitHub issue or ADO work item
- 202605-specific image/test result evidence

Please add the missing information to the PR description. Tracker details should cover day-one/regression status, affected branch/image/platform/test, and why 202605 needs the fix; validation should identify the tested 202605 image and include the result or result link.

**Comment by StormLiangMS** [comment]
Hi @bchovva,

Could you please clarify whether this PR is a **must-have cherry-pick for the 202605 branch**?

**Current evaluation**
- Classification: Documentation/metadata
- Bug fix: No
- Recommendation: **NO**
- Rationale: Adds only a 643-line T0 storm-control test-plan document, with no executable coverage or runtime correction relevant to 202605.

If this is required for 202605, please reply with the branch-specific failure or impact, required dependencies, and 202605 validation results.

If there is no response by **September 2, 2026**, the Request for 202605 branch will be rejected based on the evaluation above.
